Network/Socket/Flag.hsc

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+#include "HsNet.h"
+
+module Network.Socket.Flag where
+
+import qualified Data.Semigroup as Sem
+
+import Network.Socket.Imports
+
+-- | Message flags. To combine flags, use '(<>)'.
+newtype MsgFlag = MsgFlag { fromMsgFlag :: CInt }
+                deriving (Show, Eq, Ord, Num, Bits)
+
+instance Sem.Semigroup MsgFlag where
+    (<>) = (.|.)
+
+instance Monoid MsgFlag where
+    mempty = 0
+#if !(MIN_VERSION_base(4,11,0))
+    mappend = (Sem.<>)
+#endif
+
+-- | Send or receive OOB(out-of-bound) data.
+pattern MSG_OOB :: MsgFlag
+#ifdef MSG_OOB
+pattern MSG_OOB = MsgFlag (#const MSG_OOB)
+#else
+pattern MSG_OOB = mempty
```

Comment:
Are these correct? I though PatternSynonyms only allowed Patterns on the RHS, of which a function call isn't. I'm having trouble with some of these.

Reply:
I don't understand this.

- Q1) What does RHS mean?
- Q2) Which one is a function all? `MsgFlag`?

Reply:
1) RHS = Right Hand Side. my understanding is that pattern synonyms creates an alias for a pattern, byt `mempty` isn't a pattern.
2) yes all of them fail. using `0` instead of `mempty` works as numeric literals are patterns. You should be able to see the same error if you change `ifdef` to `ifndef`.

Reply:
I understand this and fixed the code. Nice catch!